wave_utils/include/wave/utils/config.hpp

```diff
@@ -131,93 +128,125 @@ class ConfigParser {
-    void addParam(std::string key, int *out, bool optional = false);
-    void addParam(std::string key, float *out, bool optional = false);
-    void addParam(std::string key, double *out, bool optional = false);
-    void addParam(std::string key, std::string *out, bool optional = false);
-    void addParam(std::string key,
-                  std::vector<bool> *out,
-                  bool optional = false);
-    void addParam(std::string key,
-                  std::vector<int> *out,
-                  bool optional = false);
-    void addParam(std::string key,
-                  std::vector<float> *out,
-                  bool optional = false);
-    void addParam(std::string key,
-                  std::vector<double> *out,
-                  bool optional = false);
-    void addParam(std::string key,
-                  std::vector<std::string> *out,
-                  bool optional = false);
-    void addParam(std::string key, Vec2 *out, bool optional = false);
-    void addParam(std::string key, Vec3 *out, bool optional = false);
-    void addParam(std::string key, Vec4 *out, bool optional = false);
-    void addParam(std::string key, VecX *out, bool optional = false);
-    void addParam(std::string key, Mat2 *out, bool optional = false);
-    void addParam(std::string key, Mat3 *out, bool optional = false);
-    void addParam(std::string key, Mat4 *out, bool optional = false);
-    void addParam(std::string key, MatX *out, bool optional = false);
-    void addParam(std::string key, cv::Mat *out, bool optional = false);
+    template <typename T>
+    void addParam(std::string key, T *out, bool optional = false) {
+        auto ptr = std::make_shared<ConfigParam<T>>(key, out, optional);
+        this->params.push_back(ptr);
+    }
 
     /** Get yaml node given yaml `key`. The result is assigned to `node` if
      * `key` matches anything in the config file, else `node` is set to `NULL`.
      */
-    int getYamlNode(std::string key, YAML::Node &node);
-
-    /** @return a status code meaning
-      * - `0`: On success
-      * - `-1`: Config file is not loaded
-      * - `-2`: `key` not found in yaml file, parameter is not optional
-      * - `-3`: `key` not found in yaml file, parameter is optional
-      * - `-4`: Invalid vector (wrong size)
-      * - `-5`: Invalid matrix (missing yaml key 'rows', 'cols' or 'data' for
-      * matrix)
-      */
-    /** @return see `getYamlNode` */
-    int checkKey(std::string key, bool optional);
-
-    /** @return see `checkKey`
+    ConfigStatus getYamlNode(std::string key, YAML::Node &node);
+
+    /** Check whether a key is present in the yaml file */
+    ConfigStatus checkKey(std::string key, bool optional);
+
+    /** Load yaml param primitive, array, vector or matrix. */
+    ConfigStatus loadParam(const ConfigParamBase &param);
+
+    /** Load yaml file at `config_file`. */
+    ConfigStatus load(std::string config_file);
```

Comment:
`const std::string &` maybe, I know I haven't been putting `const` everywhere but maybe we should?

Reply:
will do